src/zenml/services/service_endpoint.py

```diff
@@ -0,0 +1,172 @@
+#  Copyright (c) ZenML GmbH 2022. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+
+from typing import Any, Optional, Tuple
+
+from pydantic import Field
+
+from zenml.logger import get_logger
+from zenml.services.service_monitor import BaseServiceEndpointHealthMonitor
+from zenml.services.service_status import ServiceState, ServiceStatus
+from zenml.utils.enum_utils import StrEnum
+from zenml.utils.typed_model import BaseTypedModel
+
+logger = get_logger(__name__)
+
+
+class ServiceEndpointProtocol(StrEnum):
+    """Possible endpoint protocol values."""
+
+    TCP = "tcp"
+    HTTP = "http"
+    HTTPS = "https"
+
+
+class ServiceEndpointConfig(BaseTypedModel):
+    """Generic service endpoint configuration.
+
+    Concrete service classes should extend this class and add additional
+    attributes that they want to see reflected and use in the endpoint
+    configuration.
+
+    Attributes:
+        name: unique name for the service endpoint
+        description: description of the service endpoint
+    """
+
+    name: str = ""
+    description: str = ""
+
+
+class ServiceEndpointStatus(ServiceStatus):
+    """Status information describing the operational state of a service
+    endpoint (e.g. a HTTP/HTTPS API or generic TCP endpoint exposed by a
+    service).
+
+    Concrete service classes should extend this class and add additional
+    attributes that make up the operational state of the service endpoint.
+
+    Attributes:
+        protocol: the TCP protocol used by the service endpoint
+        hostname: the hostname where the service endpoint is accessible
+        port: the current TCP port where the service endpoint is accessible
+    """
+
+    protocol: ServiceEndpointProtocol = ServiceEndpointProtocol.TCP
+    hostname: Optional[str] = None
+    port: Optional[int] = None
+
+    @property
+    def uri(self) -> Optional[str]:
+        """Get the URI of the service endpoint.
+
+        Returns:
+            The URI of the service endpoint or None, if the service endpoint
+            operational status doesn't have the required information.
+        """
+        if not self.hostname or not self.port or not self.protocol:
+            # the service is not yet in a state in which the endpoint hostname
+            # port and protocol are known
+            return None
+
+        return f"{self.protocol.value}://{self.hostname}:{self.port}"
```

Comment:
Not sure of the tradeoffs or what is usually done here, but would it maybe be a good idea to end with a `/` for this returned value? Thinking about future use cases where people are building on top of these endpoints with query strings?